index.ts

```diff
@@ -315,6 +315,233 @@ export async function createElasticsearchMcpServer(
+        .describe("Optional index name to get shard information for"),
+    },
+    async ({ index }) => {
+      try {
+        const response = await esClient.cat.shards({
+          index,
+          format: "json",
+        });
+
+        const shardsInfo = response.map((shard) => ({
+          index: shard.index,
+          shard: shard.shard,
+          prirep: shard.prirep,
+          state: shard.state,
+          docs: shard.docs,
+          store: shard.store,
+          ip: shard.ip,
+          node: shard.node,
+        }));
+
+        const metadataFragment = {
+          type: "text" as const,
+          text: `Found ${shardsInfo.length} shards${index ? ` for index ${index}` : ""}`,
+        };
+
+        return {
+          content: [
+            metadataFragment,
+            {
+              type: "text" as const,
+              text: JSON.stringify(shardsInfo, null, 2),
+            },
+          ],
+        };
+      } catch (error) {
+        console.error(
+          `Failed to get shard information: ${
+            error instanceof Error ? error.message : String(error)
+          }`
+        );
+        return {
+          content: [
+            {
+              type: "text" as const,
+              text: `Error: ${
+                error instanceof Error ? error.message : String(error)
+              }`,
+            },
+          ],
+        };
+      }
+    }
+  );
+
+  // Tool 5: Query Profiler
+  server.tool(
+    "profile_query",
+    "Analyze query performance using Elasticsearch Profile API",
+    {
+      index: z
+        .string()
+        .trim()
+        .min(1, "Index name is required")
+        .describe("Name of the Elasticsearch index to profile"),
+
+      queryBody: z
+        .record(z.any())
+        .refine(
+          (val) => {
+            try {
+              JSON.parse(JSON.stringify(val));
+              return true;
+            } catch (e) {
+              return false;
+            }
+          },
+          {
+            message: "queryBody must be a valid Elasticsearch query DSL object",
+          }
+        )
+        .describe("Elasticsearch query DSL to profile"),
+
+      explain: z
+        .boolean()
+        .optional()
+        .default(false)
+        .describe("Whether to include explanation of how the query was executed"),
+    },
+    async ({ index, queryBody, explain }) => {
+      try {
+        const searchRequest = {
+          index,
+          body: {
+            ...queryBody,
+            profile: true,
+            explain: explain,
+          },
+        };
+
+        const response = await esClient.search(searchRequest);
+
+        if (!response.profile) {
+          return {
+            content: [
+              {
+                type: "text" as const,
+                text: "No profiling information available. Make sure the index has profiling enabled.",
+              },
+            ],
+          };
+        }
+
+        const profileInfo: {
```

Comment:
Do we really need to define this type? I would expect that elasticsearch-js client has reasonable type def for profile response?

Reply:
[7fe9331](https://github.com/elastic/mcp-server-elasticsearch/pull/28/commits/7fe9331797dd5ec61a6ed2190300e4c7a850b251)

Outdated to this commit.